### PR TITLE
Add file system watcher

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

File: nio-impl/src/main/java/org/xnio/nio/PollingFileSystemWatcher.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Polling based file system watcher service, for use on JDK6
+ *
+ * @author Stuart Douglas
+ */
+public class PollingFileSystemWatcher implements FileSystemWatcher, Runnable {
```
Comment (dmlloyd) [inline nio-impl/src/main/java/org/xnio/nio/PollingFileSystemWatcher.java]: Given that this is implementation-independent, it should probably go right into the API module as the default implementation.  Also it should be package-private.

File: nio-impl/src/main/java/org/xnio/nio/WatchServiceFileSystemWatcher.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+
+/**
+ * File system watcher service based on JDK7 {@link WatchService}. Instantiating this class will create a new thread,
+ * that will run until {@link #close()} is called.
+ * <p/>
+ *
+ * @author Stuart Douglas
+ */
+public class WatchServiceFileSystemWatcher implements FileSystemWatcher, Runnable {
```
Comment (dmlloyd) [inline nio-impl/src/main/java/org/xnio/nio/WatchServiceFileSystemWatcher.java]: Should be package-private

File: nio-impl/src/main/java/org/xnio/nio/NioXnio.java [hunk 3]
```diff
@@ -192,6 +194,21 @@ public XnioWorker createWorker(final ThreadGroup threadGroup, final OptionMap op
...
+    public FileSystemWatcher createFileSystemWatcher(String name, OptionMap options) {
+        boolean useWatchService = options.get(Options.USE_WATCH_SERVICE, true);
+        boolean daemonThread = options.get(Options.THREAD_DAEMON, true);
+        if(useWatchService) {
+            try {
+                return new WatchServiceFileSystemWatcher(name, daemonThread);
+            } catch (LinkageError e) {
+                //ignore
+            }
+        }
```
Comment (dmlloyd) [inline nio-impl/src/main/java/org/xnio/nio/NioXnio.java:204]: here you'd delegate up to super.createFileSystemWatcher()

File: api/src/main/java/org/xnio/Options.java [hunk 4]
```diff
@@ -556,4 +556,18 @@ private Options() {}
      * The number of connections to create per connection-balancing token, if connection-balancing is enabled.
      */
     public static final Option<Integer> BALANCING_CONNECTIONS = Option.simple(Options.class, "BALANCING_CONNECTIONS", Integer.class);
+
+    /**
+     * The poll interval for poll based file system watchers.
+     *
+     * Defaults to 5000ms
+     */
+    public static final Option<Integer> WATCHER_POLL_INTERVAL = Option.simple(Options.class, "WATCHER_POLL_INTERVAL", Integer.class);
```
Comment (dmlloyd) [inline api/src/main/java/org/xnio/Options.java:563]: Minor nit that newlines and so on are ignored in javadoc; the "Defaults" sentence should just go on the same line anyway, and also needs punctuation.

File: nio-impl/src/main/java/org/xnio/nio/PollingFileSystemWatcher.java (hunk 1 above)
Comment (stuartwdouglas) [inline nio-impl/src/main/java/org/xnio/nio/PollingFileSystemWatcher.java]: If I move it into the API module I can't really make it package private, and would also mean I would need to duplicate the code in FileWatchUtils.

**Pull request merged**